### PR TITLE
fix(tools): read octet-stream and xlsx urls in urlreadtool

### PR DESCRIPTION
**Pull request by joaomdmoura**
- Makes `URLReadTool` read files served as `application/octet-stream` from an extensionless URL. Content type now resolves in three steps — `Content-Type` header, then URL path extension, then the already-fetched body's magic bytes — where each step runs only if every earlier one returned nothing, and none may override an earlier answer. That ordering makes the sniff a pure widening: it can turn a refusal into a read, never a read into a different read.
- Adds an XLSX extractor. The reported file is a spreadsheet, so sniffing alone identified it as OOXML but had nowhere to send it — `URLReadTool` would have refused it even with a correct `spreadsheetml` `Content-Type`. Uses `openpyxl`, already a core `crewai` dependency, so **no new dependency**. Sheets render as CSV under a `Sheet <name>:` heading, mirroring the PDF extractor's per-page shape.
- Public surface: the agent-facing `description` gains `XLSX`, and `tool.specs.json` is regenerated (2 lines). No signature change to `safe_get_bounded` — its `(body, content_type, final_url)` 3-tuple and both callers (`url_read_tool.py`, `rag/loaders/pdf_loader.py`) are untouched, so nothing that imports it breaks.
- Preserved: a usable `Content-Type` is still authoritative and still returns `None` for `image/png`; the URL-extension fallback still runs, for both the final and the requested URL, ahead of the sniff. Both are pinned by explicit precedence tests. All 45 pre-existing tests pass **unedited**.
- Fails closed by design, and each rule is mutation-tested — the rule was deliberately broken and a test confirmed to fail. A zip is DOCX only if `word/document.xml` is in its central directory and XLSX only if `xl/workbook.xml` is (`namelist()` reads the directory, never decompresses, so a zip bomb costs nothing). Text requires a strict, whole-body UTF-8 decode with no NUL byte — a prefix decode would split a multi-byte character, and latin-1/UTF-16 are refused rather than rescued into mojibake. An empty body identifies nothing. Without the DOCX gate, the actual reported file produced `Error: Failed to read DOCX content`; with it, an honest refusal — verified against the real bytes.
- Tests: 79 passing, 34 new (121 test lines vs 77 production). Covers PDF/DOCX/XLSX/HTML/CSV/JSON sniffing, BOM and leading whitespace before an HTML prefix, xlsx CSV-quoting of embedded commas/quotes/newlines, trailing phantom-row trimming, uncalculated formulas, `.pptx`/`.vsdx`/plain-zip refusal, malformed and truncated zips, and the two precedence invariants. Only the sync `run()` path exists — `URLReadTool` does not override `_arun`, unchanged by this PR.
- No docs change: `URLReadTool` has no page under `docs/edge/`, only a changelog mention.
- Keeps this intentionally small: no `Content-Disposition` parsing, no `safe_get_bounded` signature change, no PPTX or legacy `.doc`/`.xls` extractor (PPTX would need a new `python-pptx` dependency), no reading of `response-content-*` query params, no charset-guessing rescue for non-UTF-8 text.
- Next: `DataTypes.from_content` (`rag/data_types.py:100-154`) routes this same presigned URL to `WebPageLoader` *pre-fetch* and returns raw `%PDF-1.7…` as "webpage text" with no error — a silent wrong answer that needs its own ticket. `rag/loaders/docx_loader.py:44-50` also downloads via unbounded `safe_get` with no `max_bytes` ceiling.

Verified: `ruff check lib/` and `ruff format --check lib/` clean (917 files); `mypy` clean for the changed file (the 6 pre-existing errors are in `crewai/rag/embeddings/providers/ibm/`, identical on `main`); end-to-end run against the real reported URL now returns the spreadsheet, and a local-socket run covers PDF/DOCX/CSV/HTML success plus XLSX-shaped and PNG refusal through the real `safe_get_bounded`.

Not verified: `test_search_tools.py` fails on this branch, but it fails identically (in fact more: 5 vs 4) on `main` with the branch stashed — pre-existing and unrelated. I did not test any provider/streaming path because none exists here. I have not measured resource amplification in the newly-reachable extractors; where an object store pins octet-stream and an untrusted party supplies the bytes, sniffing newly lets that uploader reach pymupdf/python-docx/openpyxl/bs4 — bounded, since a hostile server can already reach all four today by simply declaring the `Content-Type`, but naming it rather than glossing it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> URL fetching now runs sniffing and XLSX parsing on untrusted bytes (with new bounds but not uniform zip decompression limits across extractors), alongside dependency bumps that affect optional extras.
> 
> **Overview**
> **`URLReadTool`** now resolves format in order: `Content-Type`, URL extension, then **byte sniffing** on the fetched body. That unblocks presigned/share links served as `application/octet-stream` with no path extension (S3, R2, SharePoint, etc.). Sniffing is fail-closed (PDF magic, OOXML zip markers for DOCX/XLSX only, strict UTF-8 text/HTML) and cannot override an earlier decision.
> 
> **XLSX** is added: sheets are emitted as CSV under `Sheet <name>:` headings via `openpyxl`, with scan/output cell budgets and truncation messaging to limit padded-dimension DoS. Tool descriptions and **`tool.specs.json`** mention XLSX and presigned-link behavior.
> 
> **`browser_toolkit`** replaces `cast(Tag, …)` with `isinstance` checks when parsing links.
> 
> **Dependency/security pins** are tightened in workspace and **`crewai-tools`** extras: **`gitpython>=3.1.59`**, **`snowflake-sqlalchemy>=1.11.0`**, **`beautifulsoup4>=4.13.4,<5`**, and a Python-version-split **`unstructured`** floor for the `[xml]` extra (0.24+ on 3.11+). Extensive new **`url_read_tool`** tests cover sniffing, XLSX edge cases, and precedence.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 90e8ab8fbc97d6794109de218b1afab307f08e89. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---

**Bot review triage (4 findings, all with a stated position):**

- *Bugbot — XLSX padded dimensions* → **fixed** (`ea705090f`). Reproduced worse than reported: a 4.8 KB workbook with one cell at `B100000` produced 100,000 rows / 200,000 cells, and the trailing-only trim removed none of them because the stray cell keeps the last row non-empty. Rows now skip as they stream; `_XLSX_MAX_CELLS` bounds output with a visible truncation notice.
- *CodeRabbit — ZIP matching both OOXML markers* → **fixed** (`ea705090f`). Two identities is not a positive identification; refused.
- *CodeRabbit — whitespace-only cell values* → **fixed** (`4ab2cdd64`). Padding arrives as `None` → `""` exactly, so testing exactly-empty keeps the amplification fix and stops deleting author-entered spaces; `rstrip("\n")` instead of `rstrip()` preserves a trailing space in the final cell. Hoisted out of the f-string — a backslash in an f-string expression is a SyntaxError on 3.10/3.11.
- *CodeRabbit — XLSX decompression DoS (Major)* → **partially fixed, remainder deferred.** The sheet-expansion half is fixed above. ZIP decompression-ratio limits are **not** in this PR: `.docx` already reached `python-docx` and PDFs already reached `pymupdf` on `main`, so bounding only `openpyxl` would be theatre. **Needs a Linear ticket** for a uniform decompression bound across all three extractors.

**CI note:** `tests` initially failed on `tests/tracing/test_trace_enable_disable.py` (VCR cassette miss → `ConnectionError`). Re-running the failed jobs on identical code passed on 3.10/3.11/3.12/3.13, and CI's exact shard (`cd lib/crewai && pytest --splits 8 --group 5`) passes locally (662 passed). That test imports only `crewai`/`pytest`/`tests.utils` — no `crewai_tools` — and this PR touches only `lib/crewai-tools`. `pip-audit` failures are on `snowflake-sqlalchemy`/`unstructured`; `git diff main..HEAD -- '**/pyproject.toml' uv.lock` is empty, so this PR changes no dependency.

### CODE REVIEW

File: lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py [hunk 1]
```diff
@@ -281,6 +357,47 @@ def _extract_docx(body: bytes) -> str:
...
+            sheets = []
+            for worksheet in workbook.worksheets:
+                rows = [
+                    ["" if value is None else str(value) for value in row]
+                    for row in worksheet.iter_rows(values_only=True)
+                ]
+                # Excel reports a sheet's dimension generously, so trailing
+                # phantom rows are routine and would be pages of empty commas.
+                while rows and not any(cell.strip() for cell in rows[-1]):
+                    rows.pop()
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py]: Fixed in ea705090f — and it was worse than described.

Reproduced: a 4.8 KB workbook with one stray cell at B100000 yields `dimensions='A1:B100000'`, 100,000 materialized rows, 200,000 cells. The trailing-only trim was no help at all here, because the stray cell sits in the *last* row and keeps it non-empty, so all 99,998 padding rows survived into the output.

Rows are now skipped as they stream rather than materialized and trimmed afterwards, and a cell budget (`_XLSX_MAX_CELLS`) bounds what one workbook can hand an agent. The cap is announced in the output rather than applied silently. Regression test `test_one_far_down_cell_does_not_pad_the_output` asserts 3 output lines for that exact file; mutation-verified by reverting the skip and confirming it fails.

File: lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py [hunk 2]
```diff
@@ -201,18 +221,73 @@ def _classify(media_type: str) -> str | None:
...
+                    # zip bomb costs nothing here. Naming the part that must be
+                    # present is what keeps a .pptx from reaching python-docx
+                    # and surfacing as a misleading "failed to read DOCX"
+                    # instead of an honest refusal.
+                    names = set(archive.namelist())
+            except zipfile.BadZipFile:
+                return None
+            if _DOCX_ZIP_ENTRY in names:
+                return "docx"
+            return "xlsx" if _XLSX_ZIP_ENTRY in names else None
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py]: Fixed in ea705090f. Agreed on the principle — two identities is not a positive identification, and this module is explicitly fail-closed elsewhere.

Implemented slightly more compactly than the suggestion, since "both" and "neither" collapse to the same outcome:

```python
is_docx = _DOCX_ZIP_ENTRY in names
if is_docx == (_XLSX_ZIP_ENTRY in names):
    return None
return "docx" if is_docx else "xlsx"
```

Regression test `test_zip_claiming_to_be_both_docx_and_xlsx_is_refused` added, and it asserts the refusal message specifically rather than just a non-`docx` result. Mutation-verified.

File: lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py]: Fixed in 4ab2cdd64. Both sites were real; I verified each against openpyxl before changing them.

Row dropping: padding cells arrive as `None` and normalize to `""` exactly, so `not any(values)` still removes 100% of the padding while preserving a row whose cells the author genuinely filled with spaces. That is strictly better than the `.strip()` test it replaces — same amplification fix, no data loss.

Trailing cell: confirmed `"x,value \n".rstrip()` → `"x,value"` versus `.rstrip("\n")` → `"x,value "`. Hoisted out of the f-string rather than inlined, because a backslash inside an f-string expression is a SyntaxError on 3.10/3.11 and this repo tests 3.10–3.13; `ast.parse(..., feature_version=(3, 10))` now guards it.

Tests `test_xlsx_whitespace_only_values_survive` and `test_xlsx_trailing_space_in_the_final_cell_survives` pin both.

File: lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py [hunk 3]
```diff
@@ -281,6 +357,47 @@ def _extract_docx(body: bytes) -> str:
...
+        # read_only streams the sheets rather than building the whole object
+        # graph, and data_only takes cached values so formulas do not come
+        # back as "=SUM(A1:A9)". Both matter for a workbook off an untrusted URL.
+        workbook = load_workbook(BytesIO(body), read_only=True, data_only=True)
+        try:
+            sheets = []
+            for worksheet in workbook.worksheets:
+                rows = [
+                    ["" if value is None else str(value) for value in row]
+                    for row in worksheet.iter_rows(values_only=True)
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py]: Split verdict: partially fixed, remainder deferred with a reason.

**Fixed** — the sheet-expansion half, in ea705090f. This was the sharp edge and it was measurable: a 4.8 KB workbook expanded to 200,000 cells purely through `openpyxl` padding rows to the declared dimension. Rows now stream and are skipped rather than materialized, and `_XLSX_MAX_CELLS` bounds total emitted cells with a visible truncation notice. Output size is therefore bounded regardless of what the sheet declares.

**Rejected as scoped to this PR** — ZIP member/decompression-ratio limits. Not because the risk is unreal, but because it is not introduced here and this is the wrong place to fix it. `.docx` URLs already reached `Document(BytesIO(body))` on `main` (`url_read_tool.py:48`, `:277`, `:306` at `main`), and PDFs already reach `pymupdf.open`. A compressed-bomb defense that covers `openpyxl` but leaves `python-docx` and `pymupdf` open would be security theatre. The correct fix is a uniform decompression bound applied to every extractor, or at the `safe_get_bounded` boundary — a separate, larger change with its own test matrix.

I am not closing that as "wontfix": it needs a Linear ticket covering all three extractors, and I have flagged it in the PR body rather than leaving this thread as the only record. Naming it explicitly in the meantime: where an object store pins `octet-stream` and an untrusted party supplies the bytes, sniffing lets that uploader select the parser — bounded by the fact that a hostile server already reaches all three today simply by declaring the `Content-Type`, which the header branch honours first.

On "stream CSV output": rows now stream into the per-sheet buffer rather than being built as a list first, and the cell budget caps the buffer. Full incremental streaming to the caller is not possible here — `_run` returns a single `str`, and `_window` slices it afterwards.

File: lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py [hunk 4]
```diff
@@ -281,6 +363,62 @@ def _extract_docx(body: bytes) -> str:
...
+            sheets = []
+            remaining = _XLSX_MAX_CELLS
+            truncated = False
+            for worksheet in workbook.worksheets:
+                # csv rather than a join: a cell may itself hold a comma, a
+                # quote or a newline, any of which would corrupt the grid.
+                buffer = StringIO()
+                writer = csv.writer(buffer, lineterminator="\n")
+                for row in worksheet.iter_rows(values_only=True):
+                    values = ["" if value is None else str(value) for value in row]
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai-tools/src/crewai_tools/tools/url_read_tool/url_read_tool.py:420]: Fixed in d29c32d8e. Confirmed, and the impact is larger than the "quick win" label — though the stated mechanism is not quite the one that fires.

**Where the description is off:** a forged `<dimension ref="A1:XFD1048576"/>` alone does *not* produce 1,048,576 rows. In read-only mode `iter_rows()` walks the `<row>` elements actually present and gap-fills between the first and last real one; it does not fabricate the row count from the declared dimension. Your own repro script confirms this — it asserts on `next(rows)` twice and stops. I ran it: that file yields **one** row of 16,384, costing 16,384 normalizations in 0.0s.

**What actually fires** is the forged width combined with a real far-down cell, which supplies the row count the dimension cannot:

```
A1="header", B100000="stray", dimension forged to A1:XFD1048576
file size:  4,848 bytes
max_row=1048576  max_col=16384
scanned:    1,638,400,000 cells in 15.2s, budget untouched at 200,000
```

So the finding stands and the root cause is exactly as you diagnosed — the budget was charged after normalization and after the blank filter, and every one of those 100,000 padded rows is blank, so it skipped for free.

**Fix:** a second, separate ceiling charged per row *before* the row is touched, because the two concerns are genuinely different — how much work a hostile sheet can demand, versus how much output an agent receives:

```python
if len(row) > scannable:
    truncated = True
    break
scannable -= len(row)
```

Regression test `test_forged_sheet_dimension_is_bounded_by_the_scan_budget` builds that exact file (asserting it stays under 10 KB) and fails if the read takes over 5s. Mutation-verified: removing the per-row charge takes it to 26s. Truncation remains announced in the output, never silent.

Note this is the sheet-expansion half only. ZIP decompression-ratio limits remain deferred for the reason in the thread above — `python-docx` and `pymupdf` were already reachable on `main` and need the same bound.

File: pyproject.toml [hunk 5]
```diff
@@ -275,6 +287,8 @@ override-dependencies = [
     "h2>=4.4.1",
     "torch>=2.13.0",
     "snowflake-connector-python>=4.7.1",
+    "snowflake-sqlalchemy>=1.11.0",
+    "unstructured[local-inference, all-docs]>=0.24.0; python_version >= '3.11'",
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply pyproject.toml]: Confirmed and fixed in 0df741968. This was a regression I introduced, and your reading of the mechanism is exactly right.

Verified from the lock before the fix: the only `unstructured` entry was

```
{ name = "unstructured", extra = ["all-docs", "local-inference"], marker = "python_full_version >= '3.11'" }
```

one version (0.24.1), no 3.10 branch. So `crewai-tools[xml]` on 3.10 would have installed no `unstructured` at all, and my rationale comment claiming 3.10 "stays on the 0.18 line" was simply false.

Root cause is the one you named: `override-dependencies` replace the whole requirement *including its marker*, so a marker on the override subtracts from the dependency set rather than partitioning it. The fix is to stop expressing the split as an override — the floor now lives in `lib/crewai-tools/pyproject.toml` as two complementary entries, and the root override for `unstructured` is gone:

```
"unstructured[local-inference, all-docs]>=0.24.0; python_version >= '3.11'",
"unstructured[local-inference, all-docs]>=0.17.2; python_version < '3.11'",
```

The lock now carries both 0.18.32 and 0.24.1 under complementary markers, confirmed:

```
{ name = "unstructured", ..., marker = "python_full_version >= '3.11' and extra == 'xml'", specifier = ">=0.24.0" }
{ name = "unstructured", ..., marker = "python_full_version < '3.11' and extra == 'xml'", specifier = ">=0.17.2" }
```

Stated plainly since it is the residual: Python 3.10 stays on the vulnerable 0.18 line, because the patched release requires 3.11. That is now a documented trade with a TODO rather than a silently missing dependency.

[the remark this answers is not in the document]
Comment (joaomdmoura) [reply pyproject.toml]: Fixed in 0df741968. You are right, and the precedent you are pointing at is already in that same file — the `nltk` entry says "Declared here, not only as a uv override, so consumers installing crewai-tools[xml] get this floor." I had noted that limitation to myself and then failed to act on it, which is worse than not noticing.

A `[tool.uv] override-dependencies` entry shapes only this workspace's lock. It does not appear in the published wheel metadata, so anyone doing `pip install crewai-tools[snowflake]` was still resolving `snowflake-sqlalchemy>=1.7.3` and could land on the vulnerable 1.10.0. Same for `[github]` and gitpython.

Raised in `lib/crewai-tools/pyproject.toml` with the advisory named at each site:

- `snowflake-sqlalchemy>=1.7.3` → `>=1.11.0` (GHSA-8g6f-qw9x-4q6q)
- `gitpython>=3.1.58,<4` → `>=3.1.59,<4` (PYSEC-2026-3785 … -3788)
- `unstructured` → marker-split floor, see the adjacent thread

The root overrides stay for the first two, matching the existing belt-and-braces pattern. `pip-audit` with CI's exact flags and no added ignores: `No known vulnerabilities found, 5 ignored`.

**Review by Vidit-Ostwal — APPROVED** [review]
LGTM
